apis/v1alpha1/adminnetworkpolicy_types.go

```diff
@@ -147,28 +169,127 @@ type AdminNetworkPolicyEgressRule struct {
+// traffic". For "Deny" and "Pass" rules, "fail closed" means: "treat the rule
+// as a 'Deny all' rule".
+//
+// +kubebuilder:validation:MaxProperties=1
+// +kubebuilder:validation:MinProperties=1
+type AdminNetworkPolicyEgressPeer struct {
+	// Namespaces defines a way to select all pods within a set of Namespaces.
+	// Note that host-networked pods are not included in this type of peer.
+	//
+	// +optional
+	Namespaces *metav1.LabelSelector `json:"namespaces,omitempty"`
+	// Pods defines a way to select a set of pods in
+	// a set of namespaces. Note that host-networked pods
+	// are not included in this type of peer.
+	//
+	// +optional
+	Pods *NamespacedPod `json:"pods,omitempty"`
+	// Nodes defines a way to select a set of nodes in
+	// the cluster (based on the node's labels). It selects
+	// the nodeIPs as the peer type by matching on the IPs
+	// present in the node.Status.Addresses field of the node.
+	// This field follows standard label selector
+	// semantics; if present but empty, it selects all Nodes.
+	//
+	// <network-policy-api:experimental>
+	// +optional
+	Nodes *metav1.LabelSelector `json:"nodes,omitempty"`
+	// Networks defines a way to select peers via CIDR blocks.
+	// This is intended for representing entities that live outside the cluster,
+	// which can't be selected by pods, namespaces and nodes peers, but note
+	// that cluster-internal traffic will be checked against the rule as
+	// well. So if you Allow or Deny traffic to `"0.0.0.0/0"`, that will allow
+	// or deny all IPv4 pod-to-pod traffic as well. If you don't want that,
+	// add a rule that Passes all pod traffic before the Networks rule.
+	//
+	// Each item in Networks should be provided in the CIDR format and should be
+	// IPv4 or IPv6, for example "10.0.0.0/8" or "fd00::/8".
+	//
+	// Networks can have upto 25 CIDRs specified.
+	//
+	// +optional
+	// +listType=set
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=25
+	Networks []CIDR `json:"networks,omitempty"`
+
+	// DomainNames provides a way to specify domain names as peers.
+	//
+	// DomainNames is only supported for ALLOW rules. In order to control
+	// access, DomainNames ALLOW rules should be used with a lower priority
+	// egress deny -- this allows the admin to maintain an explicit "allowlist"
+	// of reachable domains.
+	//
+	// DomainNames can have up to 25 domain names specified in one rule.
+	//
+	// <network-policy-api:experimental>
+	// +optional
+	// +listType=set
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=25
+	DomainNames []DomainName `json:"domainNames,omitempty"`
+}
+
+// DomainName describes one or more domain names to be used as a peer.
+//
+// DomainName can be an exact match, or use the wildcard specifier '*' to match
+// one or more labels.
+//
+// '*', the wildcard specifier, matches one or more entire labels. It does not
+// support partial matches. '*' may only be specified as a prefix.
```

Comment:
`*` is allowed based on our survey of what other DNS-based-firewall implementations do (see the NPEP).

(It also has the side effect of forcing the implementation to implement this by snooping pod DNS requests, rather than trying to implement it by pre-resolving DNS names to IP addresses, which both GKE and OpenShift did in the past and found to be unworkable.)